lib/evaluate/tools.h

```diff
@@ -88,9 +95,16 @@ template<typename A> bool IsAssumedRank(const Designator<A> &designator) {
     return false;
   }
 }
-template<typename A> bool IsAssumedRank(const Expr<A> &expr) {
+template<typename T> bool IsAssumedRank(const Expr<T> &expr) {
   return std::visit([](const auto &x) { return IsAssumedRank(x); }, expr.u);
 }
+template<typename A> bool IsAssumedRank(const std::optional<A> &x) {
+  if (x.has_value()) {
+    return IsAssumedRank(*x);
+  } else {
+    return false;
+  }
```

Comment:
This is just personal preference, but I would write this as `return x.has_value() && IsAssumedRank(*x)`.

Reply:
I would prefer that myself, but I've been trying to be more verbose for long-term readability; I don't really like using `if (x != nullptr)` over `if (x)` either, but maybe it helps somebody in the future understand the code better.

Why wouldn't you use `return x && IsAssumedRank(*x);`?

Or maybe `IsAssumedRank()` should handle `std::optional<>` arguments.  I'll try that.

Reply:
I guess in this case I don't find the more verbose form to be more readable.

I'm fine with `return x && IsAssumedRank(*x);` too.

Reply:
Whatever you like.  I'll change it here and elsewhere.